text/0517-io-os-reform.md

```diff
@@ -473,12 +650,346 @@ throughout IO, we can go on to explore the modules in detail.
+
+#### `Errors`
+[Errors]: #error
+
+The `IoError` type will be renamed to `std::io::Error`, following our
+[non-prefixing convention](https://github.com/rust-lang/rfcs/pull/356).
+It will remain largely as it is today, but its fields will be made
+private. It may eventually grow a field to track the underlying OS
+error code.
+
+The `std::io::IoErrorKind` type will become `std::io::ErrorKind`, and
+`ShortWrite` will be dropped (it is no longer needed with the new
+`Write` semantics), which should decrease its footprint. The
+`OtherIoError` variant will become `Other` now that `enum`s are
+namespaced. Other variants may be added over time, such as `Interrupted`,
+as more errors are classified from the system.
+
+The `EndOfFile` variant will be removed in favor of returning `Ok(0)`
+from `read` on end of file (or `write` on an empty slice for example). This
+approach clarifies the meaning of the return value of `read`, matches Posix
+APIs, and makes it easier to use `try!` in the case that a "real" error should
+be bubbled out. (The main downside is that higher-level operations that might
+use `Result<T, IoError>` with some `T != usize` may need to wrap `IoError` in a
+further enum if they wish to forward unexpected EOF.)
+
+#### Channel adapters
+[Channel adapters]: #channel-adapters
+
+The `ChanReader` and `ChanWriter` adapters will be left as they are today, and
+they will remain `#[unstable]`. The channel adapters currently suffer from a few
+problems today, some of which are inherent to the design:
+
+* Construction is somewhat unergonomic. First a `mpsc` channel pair must be
+  created and then each half of the reader/writer needs to be created.
+* Each call to `write` involves moving memory onto the heap to be sent, which
+  isn't necessarily efficient.
+* The design of `std::sync::mpsc` allows for growing more channels in the
+  future, but it's unclear if we'll want to continue to provide a reader/writer
+  adapter for each channel we add to `std::sync`.
+
+These types generally feel as if they're from a different era of Rust (which
+they are!) and may take some time to fit into the current standard library. They
+can be reconsidered for stabilization after the dust settles from the I/O
+redesign as well as the recent `std::sync` redesign. At this time, however, this
+RFC recommends they remain unstable.
+
+#### `stdin`, `stdout`, `stderr`
+[stdin, stdout, stderr]: #stdin-stdout-stderr
+
+Finally, `std::io` will provide a `stdin` reader and `stdout` and
+`stderr` writers. These will largely work as they do today, except
+that we will hew more closely to the traditional setup:
+
+* `stderr` will be unbuffered and `stderr_raw` will therefore be dropped.
+* `stdout` will be line-buffered for TTY, fully buffered otherwise.
+* most TTY functionality in `StdReader` and `StdWriter` will be moved
+   to `os::unix`, since it's not yet implemented on Windows.
+* `stdout_raw` and `stderr_raw` will be removed.
```

Comment:
Do you think that a decision on this question would affect the API design somewhat? I would expect this to largely be an implementation detail instead of altering how the API works.